python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -260,6 +267,56 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    @classmethod
+    def prepare_for_save(cls, containers):
+        cls.changed_devices = cls._detect_version_changes(containers)
+        return super(Device, cls).prepare_for_save(containers)
+
+    @classmethod
+    def _detect_version_changes(cls, containers):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Returns a dictionary of the device and which of the versions changed.
+        """
```

Comment:
Using class methods on `Shadow` classes is a good practice when it simplifies bulk operations. However, when those bulk operations amount to looping over each instance and doing the same operation on each of them, it makes much more sense to implement the functionality as instance methods (each method only needs to look at its own data, not the data of other instances).

Advantages of the latter approach:
* Looping over each instance is already taken care of by the outer storage framework
* The methods become simpler
* Unit testing becomes simpler, as you don't need set up large data structures and integrate multiple running parts (effectively coming closer to an integration test)

Even when posting events, this could be done in isolation for each instance.
However, there is no corresponding `cleanup_after_save` hook for instances. Some existing `Shadow` classes work around this by overriding the `save()` method and adding its own handler code before or after a call to `super()`.

Reply:
Sounds like "cleanup_after_save" ought to exist, and the existing `save()`s updated. In a PR of its own maybe?

Reply:
#2421 suggestion for adding cleanup_after_save per container